src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -122,16 +127,44 @@ public boolean isRunning() {
      * Runs verifyIotCertificate useCase for all the registered client certificate PEMs.
      */
     @Override
-    public void run() {
+    public synchronized void run() {
         if (isNetworkDown()) {
             logger.debug("Network is down - not refreshing certificates");
             return;
         }
 
+        if (!canRun()) {
+            logger.info("Certificates were refreshed recently - retrying later");
+            return;
+        }
+
+        nextRun.set(Instant.now().plus(Duration.ofSeconds(DEFAULT_INTERVAL_SECONDS)));
         Optional<ThingAssociations> associations = getThingsAssociatedWithCoreDevice();
         associations.ifPresent(this::refresh);
     }
 
+
+    /**
+     * Handler to react to network changes.
+     * @param connectionState - A network state
+     */
+    @Override
+    public void accept(NetworkState.ConnectionState connectionState) {
+        if (connectionState == NetworkState.ConnectionState.NETWORK_UP) {
+            logger.atInfo("Network back up - refreshing certificates");
+            run();
+        }
+    }
+
+    private boolean canRun() {
+        if (nextRun.get() == null) {
+            return true;
+        }
+
+        Instant now = Instant.now();
+        return now.compareTo(nextRun.get()) >= 0;
```

Comment:
use isbefore or isafter instead of compareto to be more clear about the intention.

Reply:
Well I want a `greaterThanOrEqual` behaviour. It is either that of comparing if it is after or equal like  `now.equals(nextRun) || now.isAfter(nextRun)`

Reply:
Yup, the only problem is how it does the equality comparison of instant. I could move it to a method like `isGreatherThanOrEquals(now, nextRun)`. Thoughts?

```java
Instant now = Instant.now();
now.equals(Instant.ofEpochMilli(now.toEpochMilli())); // Returns false
```

Reply:
Why is equality a problem? Just use `.equals`, it does the right thing.

Reply:
```
public boolean equals(Object otherInstant) {
        if (this == otherInstant) {
            return true;
        }
        if (otherInstant instanceof Instant) {
            Instant other = (Instant) otherInstant;
            return this.seconds == other.seconds &&
                   this.nanos == other.nanos;
        }
        return false;
    }
```

Reply:
You can't convert to millis and back again, it loses information. Instants contain nanoseconds too, so that granularity of data will be lost, which is why you're seeing that behavior.

If you've converted from millis, then you need to also convert what you're comparing from to millis and back to round off the nanoseconds.